Do not expose Linux kernel version to page content

bershanskiy opened the pull request:
Fixes https://github.com/brave/brave-browser/issues/33503.

Prior to this, Brave was reporting Linux kernel version. Per specification, Brave should return an empty string when site queries platform version on Linux.

Source: https://wicg.github.io/ua-client-hints/#get-the-platform-version

<!-- Add brave-browser issue below that this PR will resolve -->
Resolves 

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Open https://browserleaks.com/client-hints and ensure that `platformVersion` and `Sec-CH-UA-Platform-Version` say `empty` and `""`, respectively. Also, in the console run `navigator.userAgentData.getHighEntropyValues(['platformVersion']).then(a => console.log(a.platformVersion === ''))` and ensure that `true` is printed out.

bershanskiy commented on patches/components-embedder_support-user_agent_utils.cc.patch:
```diff
@@ -10,3 +10,12 @@ index c957a691efcac5c4574fae032a876d97ad60fafc..196bb8cdf26b0f5395a33e520d63bf90
+ #if BUILDFLAG(IS_WIN)
+   metadata.platform_version = GetWindowsPlatformVersion();
++#elif BUILDFLAG(IS_LINUX)
++  metadata.platform_version = "";
```
Thanks for the review. Unfortunately, I believe such trick with macros is not viable in some environments. In particular, I get the following error:
```
> autoninja -C /home/anton/Desktop/brave/pc/src/out/Component brave -k 1 --offline
ninja: Entering directory `/home/anton/Desktop/brave/pc/src/out/Component'
[1/5] CXX obj/components/embedder_support/browser_util/user_agent_utils.o
In file included from ../../brave/chromium_src/components/embedder_support/user_agent_utils.cc:24:
../../../src/components/embedder_support/user_agent_utils.cc:442:10: error: no matching function for call to 'GetUserAgentMetadata_ChromiumImpl'
  442 |   return GetUserAgentMetadata(nullptr, only_low_entropy_ch);
      |          ^~~~~~~~~~~~~~~~~~~~
../../brave/chromium_src/components/embedder_support/user_agent_utils.cc:22:30: note: expanded from macro 'GetUserAgentMetadata'
   22 | #define GetUserAgentMetadata GetUserAgentMetadata_ChromiumImpl
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/components/embedder_support/user_agent_utils.cc:441:26: note: candidate function not viable: requires single argument 'only_low_entropy_ch', but 2 arguments were provided
  441 | blink::UserAgentMetadata GetUserAgentMetadata(bool only_low_entropy_ch) {
      |                          ^                    ~~~~~~~~~~~~~~~~~~~~~~~~
../../brave/chromium_src/components/embedder_support/user_agent_utils.cc:22:30: note: expanded from macro 'GetUserAgentMetadata'
   22 | #define GetUserAgentMetadata GetUserAgentMetadata_ChromiumImpl
      |                              ^
```

This is because the first implementation `GetUserAgentMetadata` refers to the second one. Specifically, the first implementation takes one argument gets parsed prior to the second definition which takes two, and at the time it is parsed and token is substituted, the second one is not substituted yet. If I change the order of functions (cut the first one and paste it after the second one), the error goes away. I'll paste the diff below, but I do not think it is a viable solution.

The 3-line patch I posted earlier is likely the cleanest solution.

bershanskiy replied:
```diff
diff --git a/chromium_src/components/embedder_support/user_agent_utils.cc b/chromium_src/components/embedder_support/user_agent_utils.cc
index 5681ef02c1..0ebe5ffed0 100644
--- a/chromium_src/components/embedder_support/user_agent_utils.cc
+++ b/chromium_src/components/embedder_support/user_agent_utils.cc
@@ -19,6 +19,21 @@ constexpr char kBraveBrandNameForCHUA[] = "Brave";
 // constructing the UA in brave/browser/brave_content_browser_client.cc, but we
 // can't use it here in the //components.
 #define BRAVE_GET_USER_AGENT_BRAND_LIST brand = kBraveBrandNameForCHUA;
+#define GetUserAgentMetadata GetUserAgentMetadata_ChromiumImpl
 
 #include "src/components/embedder_support/user_agent_utils.cc"
 #undef BRAVE_GET_USER_AGENT_BRAND_LIST
+#undef GetUserAgentMetadata
+
+namespace embedder_support {
+
+blink::UserAgentMetadata GetUserAgentMetadata(const PrefService* pref_service,
+                                              bool only_low_entropy_ch) {
+  blink::UserAgentMetadata metadata = GetUserAgentMetadata_ChromiumImpl(pref_service, only_low_entropy_ch);
+#if BUILDFLAG(IS_LINUX)
+  metadata.platform_version = "";
+#endif
+  return metadata;
+}
+
+}
```

bershanskiy replied:
Diff to be applied to `brave-core`:
```diff
diff --git a/chromium_src/components/embedder_support/user_agent_utils.cc b/chromium_src/components/embedder_support/user_agent_utils.cc
index 5681ef02c1..0ebe5ffed0 100644
--- a/chromium_src/components/embedder_support/user_agent_utils.cc
+++ b/chromium_src/components/embedder_support/user_agent_utils.cc
@@ -19,6 +19,21 @@ constexpr char kBraveBrandNameForCHUA[] = "Brave";
 // constructing the UA in brave/browser/brave_content_browser_client.cc, but we
 // can't use it here in the //components.
 #define BRAVE_GET_USER_AGENT_BRAND_LIST brand = kBraveBrandNameForCHUA;
+#define GetUserAgentMetadata GetUserAgentMetadata_ChromiumImpl
 
 #include "src/components/embedder_support/user_agent_utils.cc"
 #undef BRAVE_GET_USER_AGENT_BRAND_LIST
+#undef GetUserAgentMetadata
+
+namespace embedder_support {
+
+blink::UserAgentMetadata GetUserAgentMetadata(const PrefService* pref_service,
+                                              bool only_low_entropy_ch) {
+  blink::UserAgentMetadata metadata = GetUserAgentMetadata_ChromiumImpl(pref_service, only_low_entropy_ch);
+#if BUILDFLAG(IS_LINUX)
+  metadata.platform_version = "";
+#endif
+  return metadata;
+}
+
+}
diff --git a/patches/components-embedder_support-user_agent_utils.cc.patch b/patches/components-embedder_support-user_agent_utils.cc.patch
index 241e474687..697708e4b9 100644
--- a/patches/components-embedder_support-user_agent_utils.cc.patch
+++ b/patches/components-embedder_support-user_agent_utils.cc.patch
@@ -10,3 +10,26 @@ index c957a691efcac5c4574fae032a876d97ad60fafc..196bb8cdf26b0f5395a33e520d63bf90
    absl::optional<std::string> maybe_brand_override =
        base::GetFieldTrialParamValueByFeature(features::kGreaseUACH,
                                               "brand_override");
+@@ -437,9 +438,11 @@ std::string GetPlatformForUAMetadata() {
+ #endif
+ }
+ 
++/*
+ blink::UserAgentMetadata GetUserAgentMetadata(bool only_low_entropy_ch) {
+   return GetUserAgentMetadata(nullptr, only_low_entropy_ch);
+ }
++*/
+ 
+ blink::UserAgentMetadata GetUserAgentMetadata(const PrefService* pref_service,
+                                               bool only_low_entropy_ch) {
+@@ -505,6 +508,10 @@ blink::UserAgentMetadata GetUserAgentMetadata(const PrefService* pref_service,
+   return metadata;
+ }
+ 
++blink::UserAgentMetadata GetUserAgentMetadata(bool only_low_entropy_ch) {
++  return GetUserAgentMetadata(nullptr, only_low_entropy_ch);
++}
++
+ #if BUILDFLAG(IS_ANDROID)
+ void SetDesktopUserAgentOverride(content::WebContents* web_contents,
+                                  const blink::UserAgentMetadata& metadata,
```

goodov replied:
I believe you just need to forward-declare `GetUserAgentMetadata_ChromiumImpl` version manually in the `.cc`.

```diff
diff --git a/chromium_src/components/embedder_support/user_agent_utils.cc b/chromium_src/components/embedder_support/user_agent_utils.cc
index 5681ef02c1..a92e6de642 100644
--- a/chromium_src/components/embedder_support/user_agent_utils.cc
+++ b/chromium_src/components/embedder_support/user_agent_utils.cc
@@ -11,6 +11,14 @@ constexpr char kBraveBrandNameForCHUA[] = "Brave";

 }  // namespace

+namespace embedder_support {
+
+blink::UserAgentMetadata GetUserAgentMetadata_ChromiumImpl(
+    const PrefService* local_state,
+    bool only_low_entropy_ch = false);
+
+}  // namespace embedder_support
+
 // Chromium uses `version_info::GetProductName()` to get the browser's "brand"
 // name, but on MacOS we use different names for different channels (adding Beta
 // or Nightly, for example). In the UA client hint, though, we want a consistent
@@ -19,6 +27,22 @@ constexpr char kBraveBrandNameForCHUA[] = "Brave";
 // constructing the UA in brave/browser/brave_content_browser_client.cc, but we
 // can't use it here in the //components.
 #define BRAVE_GET_USER_AGENT_BRAND_LIST brand = kBraveBrandNameForCHUA;
+#define GetUserAgentMetadata GetUserAgentMetadata_ChromiumImpl

 #include "src/components/embedder_support/user_agent_utils.cc"
 #undef BRAVE_GET_USER_AGENT_BRAND_LIST
+#undef GetUserAgentMetadata
+
+namespace embedder_support {
+
+blink::UserAgentMetadata GetUserAgentMetadata(const PrefService* pref_service,
+                                              bool only_low_entropy_ch) {
+  blink::UserAgentMetadata metadata =
+      GetUserAgentMetadata_ChromiumImpl(pref_service, only_low_entropy_ch);
+#if BUILDFLAG(IS_LINUX)
+  metadata.platform_version = "";
+#endif
+  return metadata;
+}
+
+}  // namespace embedder_support
```

bershanskiy replied:
Hi, thanks for the suggestion, it worked perfectly. Should I change anything else?

goodov replied:
> Hi, thanks for the suggestion, it worked perfectly. Should I change anything else?

Please add a unit-test into `brave/components/embedder_support/user_agent_utils_unittest.cc`

bershanskiy replied:
Done!

mkarolin commented on chromium_src/components/embedder_support/user_agent_utils.cc:
Is there a danger of running into webcompat issues with this change? cc: @pes10k

bershanskiy committed:
Do not expose Linux kernel version to page content

Prior to this commit, Brave reported real Linux kernel
version in Sec-CH-UA-Platform-Version header and via
navigator.userAgentData.getHighEntropyValues(['platformVersion'])
After this, Brave reports empty string "" which matches the spec:
https://wicg.github.io/ua-client-hints/#get-the-platform-version

goodov commented on components/embedder_support/user_agent_utils_unittest.cc:
```diff
@@ -56,4 +56,13 @@ TEST(UserAgentUtilsTest, UserAgentFromCommandLine) {
+  auto metadata = GetUserAgentMetadata();
+
+  EXPECT_TRUE(metadata.platform == "Linux");
+  EXPECT_TRUE(metadata.platform_version == "");
```
please use `EXPECT_EQ(...)` so we can see the values in the test output if it fails.

bershanskiy replied:
Done

goodov commented:
@bsclifton CI has passed https://github.com/brave/brave-core/pull/20539

we can merge this after @pes10k approval.
